### PR TITLE
Update rust toolchain version

### LINKED ISSUE #747: 2022 week 02: Merge changes from rustc repository
We currently develop RMC on a fork of the rust compiler. We synchronize with the upstream project once every sprint (bi-weekly). This issue is to track the merge for week 2 of 2022 (Sprint ending on 2022-01-18).

Now that we are migrating to consuming `rustc` via `rustup`, we also need to update the toolchain version used to build `rmc-compiler`.

### PR DESCRIPTION
**Pull request by celinval**
### Description of changes: 

Update toolchain version to the latest nightly version.

### Resolved issues:

Resolves #747

### Call-outs:

There has been a fixed that was pushed to rustc where they now normalize the struct-tail type when satisfying a Pointee obligation. I have not applied the fix yet, and this change keeps the same behavior for rmc. I wasn't able to reproduce the issue posted in the related PR yet. https://github.com/rust-lang/rust/pull/92248

### Testing:

* How is this change tested?

* Is this a refactor change?

### Checklist
- [ ] Each commit message has a non-empty body, explaining why the change was made
- [ ] Methods or procedures are documented
- [ ] Regression or unit tests are included, or existing tests cover the modified code
- [ ] My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: src/rmc-compiler/rustc_codegen_rmc/src/codegen/typ.rs [hunk 1]
```diff
@@ -1297,15 +1297,15 @@ impl<'tcx> GotocCtx<'tcx> {
     pub fn use_thin_pointer(&self, mir_type: Ty<'tcx>) -> bool {
         // ptr_metadata_ty is not defined on all types, the projection of an associated type
         return !self.is_unsized(mir_type)
-            || mir_type.ptr_metadata_ty(self.tcx) == self.tcx.types.unit;
+            || mir_type.ptr_metadata_ty(self.tcx, |ty| ty) == self.tcx.types.unit;
     }
     /// A pointer to the mir type should be a slice fat pointer.
     pub fn use_slice_fat_pointer(&self, mir_type: Ty<'tcx>) -> bool {
-        return mir_type.ptr_metadata_ty(self.tcx) == self.tcx.types.usize;
+        return mir_type.ptr_metadata_ty(self.tcx, |ty| ty) == self.tcx.types.usize;
```
Comment (danielsn) [inline src/rmc-compiler/rustc_codegen_rmc/src/codegen/typ.rs:1304]: We should have a TODO to normalize this?

File: src/rmc-compiler/rustc_codegen_rmc/src/codegen/rvalue.rs [hunk 2]
```diff
@@ -381,24 +381,11 @@ impl<'tcx> GotocCtx<'tcx> {
...
-                let layout = self.layout_of(t);
-                let size = layout.size.bytes_usize();
-                let box_ty = self.tcx.mk_box(t);
-                let box_ty = self.codegen_ty(box_ty);
-                let cbmc_t = self.codegen_ty(t);
-                let box_contents = BuiltinFn::Malloc
-                    .call(vec![Expr::int_constant(size, Type::size_t())], Location::none())
-                    .cast_to(cbmc_t.to_pointer());
-                self.box_value(box_contents, box_ty)
-            }
```
Comment (danielsn) [inline src/rmc-compiler/rustc_codegen_rmc/src/codegen/rvalue.rs:395]: Why this change?

Comment (celinval) [reply src/rmc-compiler/rustc_codegen_rmc/src/codegen/rvalue.rs:395]: This NullOp::Box had deprecated and replaced by ShallowInitBox. So they finally removed it.

Comment (celinval) [reply src/rmc-compiler/rustc_codegen_rmc/src/codegen/rvalue.rs:395]: https://github.com/rust-lang/rust/commit/57a4f4a6340036b18f47dd69d078e3c377438365

File: src/rmc-compiler/rustc_codegen_rmc/src/codegen/typ.rs [hunk 3]
```diff
@@ -1296,16 +1296,18 @@ impl<'tcx> GotocCtx<'tcx> {
...
     }
     /// A pointer to the mir type should be a vtable fat pointer.
     pub fn use_vtable_fat_pointer(&self, mir_type: Ty<'tcx>) -> bool {
-        let metadata = mir_type.ptr_metadata_ty(self.tcx);
+        let metadata = mir_type.ptr_metadata_ty(self.tcx, |ty| ty);
```
Comment (danielsn) [inline src/rmc-compiler/rustc_codegen_rmc/src/codegen/typ.rs:1310]: What if we made a function 
```rust
fn normalize_type(ty: Ty<'tcx>) -> Ty<'tcx> {
  ty
}
```

**Review by danielsn — APPROVED** [review]
Approved modulo discussion

**Comment by celinval** [comment]
> Approved modulo discussion

Thanks. I added the placeholder function as suggested.
